File: python/binary_tree_inorder_traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of integers
    def inorderTraversal(self, root):
        result = []
        while root != None:
            if root.left == None:
                result.append(root.val)
                prev = root
                root = root.right
            else:
                prev = root.left
                while prev.right != None and prev.right != root:
                    prev = prev.right
                if prev.right == None:
                    prev.right = root
                    root = root.left
                elif prev.right == root:
                    prev.right = None
                    result.append(root.val)
                    root = root.right
        return result

File: cpp/binary_tree_inorder_traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> inorderTraversal(TreeNode *root) {
        vector<int> result;
        TreeNode *prev;
        while (root != NULL)
        {
            if (root->left == NULL)
            {
                result.push_back(root->val);
                root = root->right;
            }
            else
            {
                prev = root->left;
                while (prev->right != NULL && prev->right != root)
                    prev = prev->right;
                if (prev->right == NULL)
                {
                    prev->right = root;
                    root = root->left;
                }
                else if (prev->right == root)
                {
                    result.push_back(root->val);
                    prev->right = NULL;
                    prev = root;
                    root = root->right;
                }
            }
        }
        return result;
    }
};